### PR TITLE
GAWB-23: Endpoints for Importing and Exporting Workspace Attributes From TSVs

### PR DESCRIPTION
**Pull request by ansingh7115**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -1087,6 +1087,28 @@ paths:
...
+      parameters:
+        - $ref: '#/parameters/workspaceNamespaceParam'
+        - $ref: '#/parameters/workspaceNameParam'
+      responses:
+        200:
+          description: Workspace attributes in TSV format
+          schema:
+            type: file
+        404:
+          description: Workspace or entity type does not exist
```
Comment (helgridly) [inline src/main/resources/swagger/api-docs.yaml]: nitpick: just workspace

File: src/main/resources/swagger/api-docs.yaml [hunk 2]
```diff
@@ -1729,6 +1779,30 @@ paths:
...
+      parameters:
+         - $ref: '#/parameters/workspaceNamespaceParam'
+         - $ref: '#/parameters/workspaceNameParam'
+      responses:
+        200:
+          description: Workspace attributes in TSV format
+          schema:
+            type: file
+        404:
+          description: Workspace or entity type does not exist
```
Comment (helgridly) [inline src/main/resources/swagger/api-docs.yaml]: here too

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala [hunk 3]
```diff
@@ -44,16 +44,16 @@ class HttpRawlsDAO( implicit val system: ActorSystem, implicit val executionCont
     }
   }
 
-  override def getWorkspace(ns: String, name: String)(implicit userInfo: UserInfo): Future[RawlsWorkspaceResponse] =
+  override def getWorkspace(ns: String, name: String)(implicit userInfo: WithAccessToken): Future[RawlsWorkspaceResponse] =
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala]: should probably rename `userInfo` then (and elsewhere)

File: src/main/scala/org/broadinstitute/dsde/firecloud/model/UserInfo.scala [hunk 4]
```diff
@@ -13,10 +13,15 @@ import spray.http.OAuth2BearerToken
  * orchestration and rawls as clean as possible.
  *
  */
-case class UserInfo(userEmail: String, accessToken: OAuth2BearerToken, accessTokenExpiresIn: Long, id: String) {
+
+trait WithAccessToken { val accessToken : OAuth2BearerToken }
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/model/UserInfo.scala:20]: a comment explaining what this is about would help

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 5]
```diff
@@ -76,4 +95,40 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+      }
+    }
+
+    def importAttributesFromTSV(workspaceNamespace: String, workspaceName: String, tsvString: String): Future[PerRequestMessage] = {
+      withTSVFile(tsvString) { tsv =>
+        tsv.firstColumnHeader.split(":")(0) match {
+          case "workspace" =>
+            importWorkspaceAttributeTSV(workspaceNamespace, workspaceName, tsv)
+          case _ =>
+            Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Invalid first column header should start with \"workspace\""))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: grammar? `"Invalid TSV. First column header should start with \"workspace\""`

**Review by helgridly — COMMENTED** [review]
1. rename the userInfo variables you've changed to WithAccessTokens (to something like `userToken` I guess?)
2. a few nitpicks
3. can you add some tests: `getWorkspaceAttributeCalls` for a TSV creates what you'd expect, and calling `exportWorkspaceAttributesTSV` for a mocked-out workspace (flesh out MockRawlsDAO) gives a correct-looking TSV?

File: src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/MockRawlsDAO.scala [hunk 6]
```diff
@@ -1,7 +1,8 @@
 package org.broadinstitute.dsde.firecloud.dataaccess
 
+import akka.actor.FSM.->
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/MockRawlsDAO.scala]: was this intentional? what is this?

Comment (helgridly) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/MockRawlsDAO.scala]: 👍 otherwise!

Comment (ansingh7115) [reply src/test/scala/org/broadinstitute/dsde/firecloud/dataaccess/MockRawlsDAO.scala]: It was unused, I removed it

**Pull request merged**